language/tools/move-package/src/resolution/resolving_table.rs

```diff
@@ -0,0 +1,187 @@
+    redirection: BTreeMap<QualifiedAddress, usize>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum Assignment {
+    Assign(Option<AccountAddress>),
+    Linked(usize),
+}
+
+impl ResolvingTable {
+    /// A fresh `ResolvingTable` with no bindings.
+    pub fn new() -> ResolvingTable {
+        ResolvingTable {
+            assignments: Vec::new(),
+            redirection: BTreeMap::new(),
+        }
+    }
+
+    /// Iterates over the bindings in this table that are within `pkg`'s scope.
+    pub fn bindings(
+        &self,
+        pkg: PackageName,
+    ) -> impl Iterator<Item = (NamedAddress, &Option<AccountAddress>)> {
+        let start = (pkg, NamedAddress::from(""));
+        self.redirection
+            .range(start..)
+            .take_while(move |((scope, _), _)| *scope == pkg)
+            .map(|((_, name), ix)| (*name, self.parent(*ix)))
+    }
+
+    /// Return a reference to the address that `name` is currently bound to, if one exists, or
+    /// `None` otherwise.
+    pub fn get(&self, name: QualifiedAddress) -> Option<&AccountAddress> {
+        self.parent(*self.redirection.get(&name)?).as_ref()
+    }
+
+    /// Indicates whether there is a binding in this resolving table for `name`.  A table contains a
+    /// binding if it has previously been passed into a call to `define` or `unify` (even if it has
+    /// not been assigned a concrete numerical address).
+    pub fn contains(&self, name: QualifiedAddress) -> bool {
+        self.redirection.contains_key(&name)
+    }
+
+    /// Add the binding `name = addr` to the table and propagate it across all renamings that
+    /// transitively involve `name`.  Fails if this introduces a contradiction (A path through
+    /// bindings between two account addresses that are unequal to each other), and succeeds
+    /// otherwise.
+    pub fn define(&mut self, name: QualifiedAddress, addr: Option<AccountAddress>) -> Result<()> {
+        let ix = self.get_or_create_assignment(name);
+        let Assignment::Assign(slot) = &mut self.assignments[ix] else {
+            unreachable!("Non-root assignment");
+        };
+
+        match (slot, addr) {
+            (_, None) => { /* nop */ }
+            (slot @ None, addr) => *slot = addr,
+            (Some(existing), Some(new)) => {
+                if *existing != new {
+                    return Err(unification_error(name.1, *existing, new));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Add the binding `a = b` to the table.  Fails if this introduces a contradiction (A path
+    /// through bindings between two account addresses that are unequal to each other), and succeeds
+    /// otherwise.
+    pub fn unify(&mut self, a: QualifiedAddress, b: QualifiedAddress) -> Result<()> {
+        let ix = self.get_or_create_assignment(a);
+        let jx = self.get_or_create_assignment(b);
+
+        let Some((ir, jr)) = self.get_mut_pair(ix, jx) else {
```

Comment:
While it is a neat trick. I wonder if it would just be simpler to just do this all immutably, and then borrow at the point of modification? 

Like sure... you will have to index into the vector twice, but does that ever matter perf wise? 
```
if ix == jx { return Ok(()) }
let Assignment::Assign(ia) = ir else { unreachable!("Non-root assignment"); };
let Assignment::Assign(ja) = jr else { unreachable!("Non-root assignment"); };
match (ia, ja) {
    (None, Some(_)) => self.assignment[ix] = Assignment::Linked(jx),
    (Some(_), None) | (None, None) => self.assignment[ir] = Assignment::Linked(ix),
    (Some(ia), Some(ja)) => {
        if ia != ja {
            return Err(unification_error(a.1, *ia, *ja));
        }
    }
};
```

Reply:
In other words, my worry is more about the understandability of this code long term. Took me a bit to see what was going on is all.

Reply:
You're right -- I think I got bit by the borrow checker and then tried to steer well clear of it and went too far.